pkg/cfn/manager/tasks.go

```diff
@@ -1,158 +1,208 @@
-	wg.Wait()
+	return len(t.tasks)
 }
 
-// RunSingleTask runs a task with a proper error handling
-func (c *StackCollection) RunSingleTask(t Task) []error {
-	errs := []error{}
-	appendErr := func(err error) {
-		errs = append(errs, err)
+// Describe the set
+func (t *TaskSet) Describe() string {
+	descriptions := []string{}
+	for _, task := range t.tasks {
+		descriptions = append(descriptions, task.Describe())
 	}
-	if Run(appendErr, t); len(errs) > 0 {
-		return errs
+	mode := "sequential"
+	if t.Parallel {
+		mode = "parallel"
 	}
-	return nil
+	count := len(descriptions)
+	var msg string
+	noun := "task"
+	if t.Sub {
+		noun = "sub-task"
+	}
+	switch count {
+	case 0:
+		msg = "no tasks"
+	case 1:
+		msg = fmt.Sprintf("1 %s: { %s }", noun, descriptions[0])
+		if t.Sub {
+			msg = descriptions[0] // simple description for single sub-task
+		}
+	default:
+		noun += "s"
+		msg = fmt.Sprintf("%d %s %s: { %s }", count, mode, noun, strings.Join(descriptions, ", "))
+	}
+	if t.DryRun {
+		return "(dry-run) " + msg
+	}
+	return msg
 }
 
-// CreateClusterWithNodeGroups runs all tasks required to create
-// the stacks (a cluster and one or more nodegroups); any errors
-// will be returned as a slice as soon as one of the tasks or group
-// of tasks is completed
-func (c *StackCollection) CreateClusterWithNodeGroups(onlySubset sets.String) []error {
-	if errs := c.RunSingleTask(Task{c.CreateCluster, nil}); len(errs) > 0 {
-		return errs
+// Do will run through the set in the backround, it may return an error immediately,
+// or eventually write to the errs channel; it will close the channel once all tasks
+// are completed
+func (t *TaskSet) Do(errs chan error) error {
+	if t.Len() == 0 || t.DryRun {
+		logger.Debug("no actual tasks")
+		close(errs)
+		return nil
 	}
 
-	return c.CreateAllNodeGroups(onlySubset)
+	sendErr := func(err error) {
+		errs <- err
+	}
+
+	if t.Parallel {
+		go doParallelTasks(sendErr, t.tasks)
+	} else {
+		go doSequentialTasks(sendErr, t.tasks)
+	}
+	return nil
 }
 
-// CreateAllNodeGroups runs all tasks required to create the node groups;
-// any errors will be returned as a slice as soon as one of the tasks
-// or group of tasks is completed
-func (c *StackCollection) CreateAllNodeGroups(onlySubset sets.String) []error {
+// DoAllSync will run through the set in the foregounds and return all the errors
+// in a slice
+func (t *TaskSet) DoAllSync() []error {
+	if t.Len() == 0 || t.DryRun {
+		logger.Debug("no actual tasks")
+		return nil
+	}
+
 	errs := []error{}
 	appendErr := func(err error) {
 		errs = append(errs, err)
 	}
 
-	createAllNodeGroups := []Task{}
-	for i := range c.spec.NodeGroups {
-		ng := c.spec.NodeGroups[i]
-		if onlySubset != nil && !onlySubset.Has(ng.Name) {
-			continue
-		}
-		t := Task{
-			Call: c.CreateNodeGroup,
-			Data: ng,
-		}
-		createAllNodeGroups = append(createAllNodeGroups, t)
+	if t.Parallel {
+		doParallelTasks(appendErr, t.tasks)
```

Comment:
I think there is still a race condition here: there will be several goroutines in parallel appending to the `errs` slice, which I think is not thread safe. I think `doParallelTasks()` should collect the errors through channels, then combine them and return them. (Or use a mutex, which I think may look uglier)

Reply:
I think you are right, I'm seeing misbehaviour in some of my tests. I was hoping unit tests will bring it to justice! :)
It sounds like a channel here would help, but let me just write up some unit tests first.